File: socks.py
import numpy as np
import pandas as pd
from scipy.stats import nbinom, beta

# Initialize parameters
n_picked = 11  # The number of socks we are going to pick

# Function for simulating sock picking
def simulate_sock_picking(n_picked):
    prior_mu = 30
    prior_sd = 15
    prior_size_param = -prior_mu**2 / (prior_mu - prior_sd**2)
    n_socks = nbinom.rvs(n=prior_size_param, p=prior_mu / (prior_mu + prior_size_param), size=1)[0]
    prop_pairs = beta.rvs(15, 2)
    n_pairs = round(np.floor(n_socks / 2) * prop_pairs)
    n_odd = n_socks - n_pairs * 2

    # Create socks array
    socks = np.repeat(np.arange(1, n_pairs + n_odd + 1), [2] * n_pairs + [1] * n_odd)
    np.random.shuffle(socks)

    # Pick socks
    picked_socks = np.random.choice(socks, size=min(n_picked, n_socks), replace=False)
    sock_counts = pd.Series(picked_socks).value_counts()

    # Return counts and parameters
    return {
        'unique': np.sum(sock_counts == 1),
        'pairs': np.sum(sock_counts == 2),
        'n_socks': n_socks,
        'n_pairs': n_pairs,
        'n_odd': n_odd,
        'prop_pairs': prop_pairs
    }

# Simulating the process 10000 times
simulations = [simulate_sock_picking(n_picked) for _ in range(10000)]

# Convert to DataFrame for easier handling
sock_sim_df = pd.DataFrame(simulations)

# Filter for specific case
post_samples_case1 = sock_sim_df[(sock_sim_df['unique'] == 11) & (sock_sim_df['pairs'] == 0)]

# Print the first few rows of the simulations
print(sock_sim_df.head())

# Median of the number of socks
print("number of socks: ", sock_sim_df['n_socks'].median())

# Histogram of the number of socks
sock_sim_df.hist(column='n_socks')

File: socks.r
n_picked <- 11 # The number of socks to pick out of the laundry

sock_sim <- replicate(10000, {
  # Generating a sample of the parameters from the priors
  prior_mu <- 30
  prior_sd <- 15
  prior_size <- -prior_mu^2 / (prior_mu - prior_sd^2)
  n_socks <- rnbinom(1, mu = prior_mu, size = prior_size)
  prop_pairs <- rbeta(1, shape1 = 15, shape2 = 2) 
  n_pairs <- round(floor(n_socks / 2) * prop_pairs)
  n_odd <- n_socks - n_pairs * 2
  
  # Simulating picking out n_picked socks
  socks <- rep(seq_len(n_pairs + n_odd), rep(c(2, 1), c(n_pairs, n_odd)))
  picked_socks <- sample(socks, size =  min(n_picked, n_socks))
  sock_counts <- table(picked_socks)
  
  # Returning the parameters and counts of the number of matched 
  # and unique socks among those that were picked out.
  c(unique = sum(sock_counts == 1), pairs = sum(sock_counts == 2),
    n_socks = n_socks, n_pairs = n_pairs, n_odd = n_odd, prop_pairs = prop_pairs)
})

# just translating sock_sim to get one variable per column
sock_sim <- t(sock_sim) 
head(sock_sim)

post_samples <- sock_sim[sock_sim[, "unique"] == 11 & 
                         sock_sim[, "pairs" ] == 0 , ]

numsocks <- median(post_samples[,"n_socks"])
print(paste("number of socks: ", numsocks))

hist(post_samples[,"n_socks"])

